src/mlpack/tests/test_catch_tools.hpp

```diff
@@ -22,10 +22,18 @@
     REQUIRE(std::abs((R) - (L)) <= (E) * std::abs(R))
 
 // Check the values of two matrices.
-inline void CheckMatrices(const arma::mat& a,
-                          const arma::mat& b,
+template <typename MatTypeA, typename MatTypeB,
+    typename = std::enable_if_t<arma::is_arma_type<MatTypeA>::value
+        && arma::is_arma_type<MatTypeB>::value
+        && std::is_same<typename MatTypeA::elem_type, typename MatTypeB::elem_type>::value
+        && !std::is_integral<typename MatTypeA::elem_type>::value
+        && !std::is_integral<typename MatTypeB::elem_type>::value>>
+inline void CheckMatrices(const MatTypeA& _a,
+                          const MatTypeB& _b,
                           double tolerance = 1e-5)
 {
+  arma::Mat<typename MatTypeA::elem_type> a(_a);
```

Comment:
This does incur a copy if the input is *not* an expression.  Sometimes `CheckMatrices()` is called a lot, so I think it may be worth avoiding the copy.  I wonder if it would be better to restructure something like this:

```
inline void CheckMatrices(const arma::Mat<eT>& a, const arma::Mat<eT>& b)
{
  // old implementation
}

inline void CheckMatrices(const MatTypeA& a, const MatTypeB& b,
    const std::enable_if<MatTypeA and MatTypeB are not directly matrices but instead expressions>)
{
  arma::Mat<eT> evaluatedA(a);
  arma::Mat<eT> evaluatedB(b);
  CheckMatrices(evaluatedA, evaluatedB);
}
```

Although that doesn't capture the case where one argument is an expression and the other is a matrix, it would be possible to either add more versions for those cases, or to build some simple template struct that either stores a reference to a matrix (if what's given is a matrix), or evaluates the expression and stores it in a new matrix (if what's given is not a matrix).  I dunno, either works.